0012/README.md

```diff
@@ -0,0 +1,186 @@
+	GetWorkflowContexts(ctx context.Context, in *WorkflowContextRequest, opts ...grpc.CallOption) (WorkflowSvc_GetWorkflowContextsClient, error)
+	GetWorkflowActions(ctx context.Context, in *WorkflowActionsRequest, opts ...grpc.CallOption) (*WorkflowActionList, error)
+	ReportActionStatus(ctx context.Context, in *WorkflowActionStatus, opts ...grpc.CallOption) (*Empty, error)
+	GetWorkflowData(ctx context.Context, in *GetWorkflowDataRequest, opts ...grpc.CallOption) (*GetWorkflowDataResponse, error)
+	GetWorkflowMetadata(ctx context.Context, in *GetWorkflowDataRequest, opts ...grpc.CallOption) (*GetWorkflowDataResponse, error)
+	GetWorkflowDataVersion(ctx context.Context, in *GetWorkflowDataRequest, opts ...grpc.CallOption) (*GetWorkflowDataResponse, error)
+	UpdateWorkflowData(ctx context.Context, in *UpdateWorkflowDataRequest, opts ...grpc.CallOption) (*Empty, error)
+}
+```
+
+### The New Model
+
+The new model is inspired from Kubernetes.
+The resource clients should have a `Watch` function that will stream events for that particular resource.
+
+```
+Watch(ctx context.Context, in EventWatchRequest)
+```
+
+## The API
+
+### Data Model
+
+ResourceType - a resource that an event can be associated with
+
+- Template
+- Hardware
+- Workflow
+
+EventType - an event type in tinkerbell space(a non-exhaustive list)
+- CREATED 
+- UPDATED
+- DELETED
+- WORKFLOW_STARTED
+- WORKFLOW_INPROGRESS
+- WORKFLOW_FAILED
+- WORKFLOW_TIMEOUT
+
+Event - an event in tinkerbell space; and has the following structure:
+
+```
+{
+  "id": "uuid",
+  "resourceID": "uuid",
+  "resourceType": "workflow",
+  "eventType": "created",
+  "time": "", 
+  "data": {
+  }
+}
+```
+
+- `id`: a unique identifier (UUID) for the event occured in tinkerbell space
+- `resourceID`: a unique identifier (UUID) for the resource, the event is associated with.
+For example, `resourceID` will be set to `workflowID` for workflow resource and to `templateID` for the template resource.
+- `resouceType`: type of the resource (template, hardware, workflow) for which the event was generated
+- `eventType`: the event verb; describing the action on resource that generated the event
+- `time`: the event timestamp
+- `data`: the primary event payload, represented as `interface{}`.
+For a workflow created event, the payload (data) can be the complete workflow structure.
+For an on-timeout event the data can be an action that needs to be executed next.
+
+### EventClient
+
+A new `EventClient` should be developed with the primitive required for the events at least `Watch`.
+At the beginning, only a `Watch` function is required but we think a natural evolution will be to serve a function that can be used from other components to fire events.
+Or maybe we can do a `Watch` and `Create` straight away.
+All the resource clients will have a `Watch` function that will stream events for that particular resource.
```

Comment:
It returns something that waits and runs your code.

Reply:
Watch is going to be as simple as an infinite `for` loop running in a goroutine. All the events as they occur will be stored in the events table and will be streamed over gRPC and it will be up to the consumer how they want to consume the stream.